lib/xlibc-rt/chkstk.s

```diff
@@ -0,0 +1,29 @@
+/*
+    This file is licensed under the CC0 1.0.
+    For details, see: https://creativecommons.org/publicdomain/zero/1.0/
+*/
+
+.text
+
+.globl __chkstk
+__chkstk:
+    pushl %ecx
+    leal 8(%esp), %ecx  // Load original stack address into ecx
+    pushl %ecx
+    pushl %eax
+
+    pushl %ecx
+    pushl %eax
+    call __xlibc_check_stack
+    addl $8, %esp
+
+    popl %eax
+    popl %ecx
+    subl %eax, %ecx     // ecx is the new stack pointer
+    leal 4(%esp), %eax  // eax is a pointer to the return address variable
+    movl %ecx, %esp     // load the new stack address
+    movl -4(%eax), %ecx // restore ecx
```

Comment:
What's the calling convention of this function? Do we really have to restore ecx?

(Might also be worth noting the context on entry in a comment for the function: requested size is in eax etc.)

Reply:
chkstk doesn't really comply to any standardized calling convention (for example, %esp gets modified). As far as I understood it, a call to chkstk is supposed to have the same effect as "subl %eax, %esp", so we're not allowed to modify %eax or trash other registers (actually, I forgot to also save and restore %edx, which is specified as caller-saved in cdecl, when calling _xlibc_check_stack).
I'll add a comment to the function to clarify things a bit.